src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -0,0 +1,189 @@
+  resolver_.ResolveDependencies(task_spec, [this, task_spec]() {
+    // TODO(ekl) should have a queue per distinct resource type required
+    absl::MutexLock lock(&mu_);
+    RequestNewWorkerIfNeeded(task_spec);
+    queued_tasks_.push_back(task_spec);
+    // The task is now queued and will be picked up by the next leased or newly
+    // idle worker. We are guaranteed a worker will show up since we called
+    // RequestNewWorkerIfNeeded() earlier while holding mu_.
+  });
+  return Status::OK();
+}
+
+void CoreWorkerDirectTaskSubmitter::HandleWorkerLeaseGranted(const std::string &address,
+                                                             int port) {
+  WorkerAddress addr = std::make_pair(address, port);
+
+  // Setup client state for this worker.
+  {
+    absl::MutexLock lock(&mu_);
+    worker_request_pending_ = false;
+
+    auto it = client_cache_.find(addr);
+    if (it == client_cache_.end()) {
+      client_cache_[addr] =
+          std::shared_ptr<rpc::CoreWorkerClientInterface>(client_factory_(addr));
+      RAY_LOG(INFO) << "Connected to " << address << ":" << port;
+    }
+  }
+
+  // Try to assign it work.
+  WorkerIdle(addr, /*error=*/false);
+}
+
+void CoreWorkerDirectTaskSubmitter::WorkerIdle(const WorkerAddress &addr,
+                                               bool was_error) {
+  absl::MutexLock lock(&mu_);
+  if (queued_tasks_.empty() || was_error) {
+    RAY_CHECK_OK(lease_client_.ReturnWorker(addr.second));
+  } else {
+    auto &client = *client_cache_[addr];
+    PushTask(addr, client, queued_tasks_.front());
+    queued_tasks_.pop_front();
+  }
+  // We have a queue of tasks, try to request more workers.
+  if (!queued_tasks_.empty()) {
+    RequestNewWorkerIfNeeded(queued_tasks_.front());
+  }
+}
+
+void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
+    const TaskSpecification &resource_spec) {
+  if (worker_request_pending_) {
+    return;
+  }
+  RAY_CHECK_OK(lease_client_.RequestWorkerLease(resource_spec));
+  worker_request_pending_ = true;
+}
+
+void CoreWorkerDirectTaskSubmitter::TreatTaskAsFailed(const TaskID &task_id,
+                                                      int num_returns,
+                                                      const rpc::ErrorType &error_type) {
+  RAY_LOG(DEBUG) << "Treat task as failed. task_id: " << task_id
+                 << ", error_type: " << ErrorType_Name(error_type);
+  for (int i = 0; i < num_returns; i++) {
+    const auto object_id = ObjectID::ForTaskReturn(
+        task_id, /*index=*/i + 1,
+        /*transport_type=*/static_cast<int>(TaskTransportType::DIRECT));
+    std::string meta = std::to_string(static_cast<int>(error_type));
+    auto metadata = const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(meta.data()));
+    auto meta_buffer = std::make_shared<LocalMemoryBuffer>(metadata, meta.size());
+    RAY_CHECK_OK(in_memory_store_.Put(RayObject(nullptr, meta_buffer), object_id));
```

Comment:
this function can be made common for actor task and non-actor task?

Reply:
Maybe later... there are a couple subtle differences.